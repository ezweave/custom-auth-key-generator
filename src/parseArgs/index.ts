import { Args, ArgTypes, parseArgs } from './parseArgs';

export { Args, ArgTypes, parseArgs };
