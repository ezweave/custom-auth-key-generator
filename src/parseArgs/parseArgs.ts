import { get } from 'lodash';

export enum ArgTypes {
    configFile = 'configFile',
    headerPrefix = 'headerPrefix',
    path = 'path',
    privateKey = 'privateKey',
    publicKey = 'publicKey',
}

export interface Args {
    configFile?: string
    headerPrefix?: string,
    path: string,
    privateKey?: string,
    publicKey?: string,
}

export interface ParseArgs {
// eslint-disable-next-line @typescript-eslint/no-explicit-any 
    (args: any): Args
}

export const parseArgs: ParseArgs = (
  args 
) => ({
    configFile: get(args, ArgTypes.configFile),
    headerPrefix: get(args, ArgTypes.headerPrefix),
    path: get(args, ArgTypes.path),
    privateKey: get(args, ArgTypes.privateKey),
    publicKey: get(args, ArgTypes.publicKey),
});
