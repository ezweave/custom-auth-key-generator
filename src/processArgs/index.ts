import { CompleteArgs, processArgs } from './processArgs';

export { CompleteArgs, processArgs };
