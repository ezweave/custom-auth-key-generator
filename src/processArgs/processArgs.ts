import { get, map, overEvery, reduce } from 'lodash';

import { getEnv } from '../env';
import { Args, ArgTypes } from '../parseArgs';
import { readConfigFile } from '../readConfigFile';


export interface CompleteArgs {
    headerPrefix: string,
    path: string,
    privateKey: string,
    publicKey: string,
}

export interface ProcessArgs {
    (args: Args): CompleteArgs 
}

const isFieldPresent = (
    fieldName: string
) => (
    args: Args
) => !!get(args, fieldName);

const areAllFieldsPresent = (
    fieldNames: Array<string>
) => overEvery(
    map(fieldNames, isFieldPresent)
);

/**
 * These are all the arguments required to build a key
 */
const completeArgs = [
    ArgTypes.path,
    ArgTypes.privateKey,
    ArgTypes.publicKey,
    ArgTypes.headerPrefix,
];

export const isFromArgs = areAllFieldsPresent([
    ArgTypes.privateKey,
    ArgTypes.publicKey,
    ArgTypes.headerPrefix,
]);

export const isFromConfigFile = areAllFieldsPresent([ArgTypes.configFile]); 

export const isPathPresent = areAllFieldsPresent([ArgTypes.path]);

export const areArgsComplete = areAllFieldsPresent(completeArgs);

const argCheck = (args: Args) => (missingArgs: Array<string>, argName: string): Array<string> => {
    const foundValue = get(args, argName);
    return foundValue === undefined || foundValue === 'undefined' ? [
        ...missingArgs,
        `Missing value for required argument, ${argName}`, 
    ] : missingArgs;
};

export const buildErrorArrayForMissingArgs = (args: Args): Array<string> => reduce(
        completeArgs,
        argCheck(args),
        []
    );

export const throwErrorIfArgsAreCompleteOtherwiseReturnArgs = (args: Args): CompleteArgs => {
    if(areArgsComplete(args)) {
        return args as CompleteArgs;
    } else {
        const errors = buildErrorArrayForMissingArgs(args);
        throw Error(`Missing arguments: \n\t ${errors.join('\n\t')}`);
    }
};


export const processArgs: ProcessArgs = (
    args
) => {
    /**
     * 1. path must be supplied.  If this is coming from the command line, it is enforced by yargs
     * 2. Is this from a config file?  Load the args from the config file
     * 3. If not, load the environment.
     * 4. Check that all args are present.
     * 5. Return complete args
     */
    if(isFromConfigFile(args)) {
        const config = readConfigFile(args.configFile as string);
        const argsWithConfigFromFile = {
            ...config,
            path: args.path,
        };
        return throwErrorIfArgsAreCompleteOtherwiseReturnArgs(argsWithConfigFromFile);
    }
    if(isFromArgs(args)) {
        return throwErrorIfArgsAreCompleteOtherwiseReturnArgs(args);
    } else {
        const env = getEnv();
        const argsWithEnv = {
            ...env,
            path: args.path,
        };
        return throwErrorIfArgsAreCompleteOtherwiseReturnArgs(argsWithEnv);
    }
};
