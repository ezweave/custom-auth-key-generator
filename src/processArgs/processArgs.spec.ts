import { Args } from '../parseArgs';
import { areArgsComplete, buildErrorArrayForMissingArgs, isFromArgs, isFromConfigFile, isPathPresent, processArgs, throwErrorIfArgsAreCompleteOtherwiseReturnArgs } from './processArgs';

const completeArgs = {
   headerPrefix: 'prefix',
   path: '/foo',
   privateKey: 'private',
   publicKey: 'public',
};

const partialArgs = {
    path: '/foo',
    privateKey: 'private', 
};

const badArgsNoTypeChecks = {} as unknown as Args;

describe('processArgs', () => {
   describe('isFromArgs', () => {
       it('returns true if all required args are present', () => {
           expect(
               isFromArgs(completeArgs)
           ).toEqual(true);
       });
       it('returns false if args are missing', () => {
           expect(
               isFromArgs({
                   privateKey: 'private',
                   path: '/foo',
               })
           ).toEqual(false);
       });
   });
   describe('isFromConfigFile', () => {
       it('returns true if a config file is specified', () => {
           expect(
               isFromConfigFile({
                   configFile: 'file.yaml',
                   path: '/foo',
               })
           ).toEqual(true);
       });
       it('returns false if no config file is specified', () => {
           expect(
               isFromConfigFile({
                   path: '/foo',
               })
           ).toEqual(false);
       });
   });
   describe('isPathPresent', () => {
       it('returns true if a config file is specified', () => {
           expect(
               isPathPresent({
                   configFile: 'file.yaml',
                   path: '/foo',
               })
           ).toEqual(true);
       });
       it('returns false if no config file is specified', () => {
           expect(
            isPathPresent(badArgsNoTypeChecks)
           ).toEqual(false);
       });
   });
   describe('areArgsComplete', () => {
       it('returns true if args are complete', () => {
           expect(
               areArgsComplete(completeArgs)
           ).toEqual(true);
       });
       it('returns false if args are incomplete', () => {
           expect(
             areArgsComplete(badArgsNoTypeChecks)
           ).toEqual(false);
       });
   });
   describe('buildErrorArrayForMissingArgs', () => {
        it('is empty if all args are present', () => {
            expect(
                buildErrorArrayForMissingArgs(completeArgs)
            ).toEqual([]);
        });
        it('returns an array of missing arg errors', () => {
            expect(
                buildErrorArrayForMissingArgs(partialArgs)
            ).toMatchSnapshot();
        });
   });
   describe('throwErrorIfArgsAreCompleteOtherwiseReturnArgs', () => {
       it('returns true if all args are present', () => {
           expect(
               throwErrorIfArgsAreCompleteOtherwiseReturnArgs(completeArgs)
           ).toMatchSnapshot();
       });
       it('throws an error if anything is missing', () => {
           expect(
            () => throwErrorIfArgsAreCompleteOtherwiseReturnArgs(badArgsNoTypeChecks)
           ).toThrowErrorMatchingSnapshot();
       });
   });

   it('throws an error if no path is present', () => {
     expect(
         () => processArgs(badArgsNoTypeChecks)
     ).toThrowErrorMatchingSnapshot();
   });
});
