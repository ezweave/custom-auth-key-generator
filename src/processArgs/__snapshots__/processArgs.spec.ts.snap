// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processArgs buildErrorArrayForMissingArgs returns an array of missing arg errors 1`] = `
Array [
  "Missing value for required argument, publicKey",
  "Missing value for required argument, headerPrefix",
]
`;

exports[`processArgs throwErrorIfArgsAreCompleteOtherwiseReturnArgs returns true if all args are present 1`] = `
Object {
  "headerPrefix": "prefix",
  "path": "/foo",
  "privateKey": "private",
  "publicKey": "public",
}
`;

exports[`processArgs throwErrorIfArgsAreCompleteOtherwiseReturnArgs throws an error if anything is missing 1`] = `
"Missing arguments: 
	 Missing value for required argument, path
	Missing value for required argument, privateKey
	Missing value for required argument, publicKey
	Missing value for required argument, headerPrefix"
`;

exports[`processArgs throws an error if no path is present 1`] = `
"Missing arguments: 
	 Missing value for required argument, path"
`;
